service/worker/scanner/executions/shard_validator.go

```diff
@@ -0,0 +1,326 @@
+		TotalDBRequests   int64
+		ScanStats         shardValidationReportExecutionsScanned
+		IsFailure         bool
+		Details           string
+		Error             error
+		ExecutionFailures []ExecutionValidationFailure
+	}
+
+	// ExecutionValidationFailure details of single execution validation failure.
+	ExecutionValidationFailure struct {
+		Namespace   string
+		WorkflowID  string
+		RunID       string
+		BranchID    string
+		TreeID      string
+		NextEventID int64
+		CloseStatus enumspb.WorkflowExecutionStatus
+
+		// ValidatorTag -> ValidatorFailure
+		Failures map[string]ExecutionValidatorResult
+	}
+)
+
+// CreateShardValidator returns instance of ShardValidator implementation
+func CreateShardValidator(
+	execMgrProvider persistence.ExecutionManagerFactory, limiter quotas.RateLimiter, executionsPageSize int,
+	historyDB persistence.HistoryManager, validators []ExecutionValidator,
+) ShardValidator {
+	return &shardValidatorImpl{
+		execMgrFactory: execMgrProvider, limiter: limiter, executionsPageSize: executionsPageSize,
+		historyDB: historyDB, validators: validators,
+	}
+}
+
+func (s *shardValidatorImpl) validate(shardID int32) *ShardValidationResult {
+	report := &ShardValidationResult{
+		ShardID: shardID,
+	}
+
+	execMgr, err := s.execMgrFactory.NewExecutionManager(shardID)
+	if err != nil {
+		report.IsFailure = true
+		report.Details = "Failed to receive mutableState manager"
+		report.Error = err
+		return report
+	}
+
+	var token []byte
+	isFirstIteration := true
+	for isFirstIteration || len(token) != 0 {
+		isFirstIteration = false
+		req := &persistence.ListConcreteExecutionsRequest{
+			PageSize:  executionsPageSize,
+			PageToken: token,
+		}
+		preconditionForDBCall(&report.TotalDBRequests, s.limiter)
+		resp, err := execMgr.ListConcreteExecutions(req)
+		if err != nil {
+			report.IsFailure = true
+			report.Details = "Failed to ListConreteExecutions"
+			report.Error = err
+			return report
+		}
+		token = resp.PageToken
+		for _, mutableState := range resp.States {
+			report.ScanStats.TotalExecutionsCount++
+
+			executionValidationFailure := validateExecution(
+				mutableState, shardID, s.limiter, s.historyDB, &report.TotalDBRequests, execMgr, s.validators,
+			)
+			if executionValidationFailure != nil {
+				report.IsFailure = true
+				report.Details = "One or more execution validators reported failure."
+				report.ScanStats.CorruptedExecutionsCount++
+				report.ExecutionFailures = append(report.ExecutionFailures, *executionValidationFailure)
+			}
+		}
+	}
+	return report
+}
+
+func newExecutionValidationFailure(src *ExecutionData) *ExecutionValidationFailure {
+	result := &ExecutionValidationFailure{
+		Namespace:   src.mutableState.ExecutionInfo.NamespaceId,
+		WorkflowID:  src.mutableState.ExecutionInfo.WorkflowId,
+		RunID:       src.mutableState.ExecutionState.RunId,
+		NextEventID: src.mutableState.NextEventId,
+		CloseStatus: src.mutableState.ExecutionState.Status,
+	}
+	if src.historyBranch != nil {
```

Comment:
when will the history branch be nil?

Reply:
Should not be null, but keeping this for safety check. From perspective of this method, it can be null.